warehouse/packaging/services.py

```diff
@@ -161,3 +167,70 @@ def remove_by_prefix(self, prefix):
             self.s3_client.delete_objects(
                 Bucket=self.bucket_name, Delete={"Objects": keys_to_delete}
             )
+
+
+@implementer(IFileStorage)
+class GCSFileStorage(GenericFileStorage):
+    @classmethod
+    def create_service(cls, context, request):
+        storage_client = request.find_service(name="gcloud.gcs")
+        bucket_name = request.registry.settings["files.bucket"]
+        bucket = storage_client.get_bucket(bucket_name)
+        prefix = request.registry.settings.get("files.prefix")
+
+        return cls(bucket, prefix=prefix)
+
+    def get(self, path):
+        # Note: this is not actually used in production, instead our CDN is
+        # configured to connect directly to our storage bucket. See:
+        # https://github.com/python/pypi-infra/blob/master/terraform/file-hosting/vcl/main.vcl
+        raise NotImplementedError
+
+    def store(self, path, file_path, *, meta=None):
+        path = self._get_path(path)
+        blob = self.bucket.blob(path)
+        blob.upload_from_filename(file_path)
+
+        if meta is not None:
+            blob.metadata = meta
+            blob.patch()
```

Comment:
Does this need to be a separate call or can it be added to the `blob` created above?

I can't tell from the docs but it might be worth avoiding a second request.

Reply:
As far as I can tell, this is the only way to do it.

Reply:
A quick test shows that

```python
blob = self.bucket.blob(path)
blob.metadata = meta
blob.upload_from_filename(file_path)
```

Correctly sets the metadata without having to make two calls!

Reply:
Does it work when `meta` is `None` as well?

Reply:
in the patch I'm about to send, I still do a None check :-D